lib/bosh/BoshDirectorClient.js

```diff
@@ -304,25 +304,47 @@ class BoshDirectorClient extends HttpClient {
   createOrUpdateDeployment(manifest, opts) {
     const query = opts ? _.pick(opts, 'recreate', 'skip_drain', 'context') : undefined;
     const deploymentName = yaml.safeLoad(manifest).name;
-    const config = _.sample(this.activePrimary);
-    if (config === undefined || config.length === 0) {
-      throw new errors.NotFound('Did not find any bosh director config which supports creation of deployment');
-    }
-    return this
-      .makeRequestWithConfig({
-        method: 'POST',
-        url: '/deployments',
-        headers: {
-          'Content-Type': 'text/yaml'
-        },
-        qs: query,
-        body: _.isObject(manifest) ? yaml.safeDump(manifest) : manifest
-      }, 302, config)
-      .tap(() => {
-        logger.info(`Cached ${deploymentName} at director: ${config.name} ${config.url}`);
-        this.cache[deploymentName] = config;
+    const boshDirectorName = _.get(opts, 'bosh_director_name');
+    return Promise.try(() => {
+        if (boshDirectorName) {
+          logger.warn(`Deploying through ${boshDirectorName}`);
+          return this
+            .getDirectorConfig(deploymentName)
+            .then(config => {
+              if (config) {
+                throw new BadRequest(`Cannot create an existing deployment ${deploymentName}`);
+              }
+            })
+            .catch(NotFound, () => this.getConfigByName(boshDirectorName));
+        } else {
+          return this
+            .getDirectorConfig(deploymentName) // if update
+            .catch(() => _.sample(this.activePrimary)); // if create
+        }
       })
-      .then(res => this.prefixTaskId(deploymentName, res));
+      .then(config => {
+        if (config === undefined || config.length === 0) {
+          throw new errors.NotFound('Did not find any bosh director config which supports creation of deployment');
+        }
+        return config;
```

Comment:
unnecessary promise chain. Please club the below `.then` block with this one and directly return result.

Reply:
done